File: tests/testthat/test_weather_validation.R
library(testthat)
library(dssatutils)

write_sample_wth <- function(path, rows) {
  writeLines(c(
    "$WEATHER DATA: test",
    "@ INSI      LAT     LONG  ELEV   TAV   AMP REFHT WNDHT",
    "  TEST   0.0000   0.0000   -99  10.0  20.0   2.0  10.0",
    "@  DATE  SRAD  TMAX  TMIN  RAIN  TDEW  RH2M  WIND",
    rows
  ), path)
}

test_that("fixed-width adjacent negative weather values are valid", {
  path <- tempfile(fileext = ".WTH")
  rows <- c(
    sprintf("%7s%6.1f%6.1f%6.1f%6.1f%6.1f%6.1f%6.1f", "2024001", 12, -10.2, -12.3, 0, -15, 40, 3),
    sprintf("%7s%6.1f%6.1f%6.1f%6.1f%6.1f%6.1f%6.1f", "2024002", 13, -9, -11, 0, -14, 42, 3.2)
  )
  write_sample_wth(path, rows)
  expect_true(is_wth_valid(path, end_year = 2024))
})

test_that("weather validation rejects date gaps", {
  path <- tempfile(fileext = ".WTH")
  write_sample_wth(path, c(
    "2024001 12.0 10.0 1.0 0.0 0.0 40.0 3.0",
    "2024003 12.0 10.0 1.0 0.0 0.0 40.0 3.0"
  ))
  expect_false(is_wth_valid(path, end_year = 2024))
})

test_that("weather validation rejects absolute-zero temperatures", {
  path <- tempfile(fileext = ".WTH")
  row <- sprintf("%7s%6.1f%6.1f%6.1f%6.1f%6.1f%6.1f%6.1f",
                 "2018001", 0, -273.1, -273.1, 0, -273.1, 0, 0)
  write_sample_wth(path, row)
  expect_false(is_wth_valid(path, end_year = 2018))
})

test_that("weather validation can require complete core forcing", {
  path <- tempfile(fileext = ".WTH")
  write_sample_wth(path, c(
    "2024001 -99 10.0 1.0 0.0 0.0 40.0 -99",
    "2024002 12.0 10.0 1.0 0.0 0.0 40.0 -99"
  ))
  expect_true(is_wth_valid(path, end_year = 2024))
  expect_false(is_wth_valid(
    path,
    end_year = 2024,
    required_columns = c("SRAD", "TMAX", "TMIN", "RAIN")
  ))
})

test_that("weather validation can require all AgERA5 forcing", {
  path <- tempfile(fileext = ".WTH")
  write_sample_wth(path, c(
    "2024001 12.0 10.0 1.0 0.0 -99 -99 -99",
    "2024002 12.0 10.0 1.0 0.0 -99 -99 -99"
  ))
  core <- c("SRAD", "TMAX", "TMIN", "RAIN")
  agera5 <- c(core, "TDEW", "RH2M", "WIND")

  expect_true(is_wth_valid(path, end_year = 2024, required_columns = core))
  expect_false(is_wth_valid(path, end_year = 2024, required_columns = agera5))
})

test_that("AgERA5 writer defers physical validation to the shared validator", {
  wd <- data.frame(
    DATE = c("2018001", "2018002"), YEAR = c(2018, 2018), MM = c(1, 1),
    SRAD = 12, TMAX = c(5.8, 10), TMIN = c(6, 2), RAIN = 0,
    TDEW = 0, RH2M = 60, WIND = 3
  )
  writer <- getFromNamespace(".agera5_write_wth", "dssatutils")
  path <- tempfile(fileext = ".WTH")
  out_dir <- dirname(path)
  generated <- writer(wd, tools::file_path_sans_ext(basename(path)), 33.7, -102.5, out_dir)

  expect_true(file.exists(generated))
  expect_true(any(grepl("   5.8   6.0", readLines(generated), fixed = TRUE)))
  expect_false(is_wth_valid(
    generated,
    end_year = 2018,
    required_columns = c("SRAD", "TMAX", "TMIN", "RAIN", "TDEW", "RH2M", "WIND")
  ))
})

File: tests/test_weather_validation.py
from pathlib import Path

import pandas as pd

from dssatutils import is_wth_valid
from dssatutils.weather_agera5 import _write_wth
from dssatutils.weather_nasapower import _NASA_PARAMS, _fetch_nasa_power


def _write_sample(path: Path, rows: list[str]) -> None:
    path.write_text(
        "$WEATHER DATA: test\n"
        "@ INSI      LAT     LONG  ELEV   TAV   AMP REFHT WNDHT\n"
        "  TEST   0.0000   0.0000   -99  10.0  20.0   2.0  10.0\n"
        "@  DATE  SRAD  TMAX  TMIN  RAIN  TDEW  RH2M  WIND\n" +
        "\n".join(rows) + "\n",
        encoding="utf-8",
    )


def test_fixed_width_adjacent_negative_values_are_valid(tmp_path):
    path = tmp_path / "TEST.WTH"
    rows = [
        f"{'2024001':>7}{12.0:6.1f}{-10.2:6.1f}{-12.3:6.1f}{0.0:6.1f}{-15.0:6.1f}{40.0:6.1f}{3.0:6.1f}",
        f"{'2024002':>7}{13.0:6.1f}{-9.0:6.1f}{-11.0:6.1f}{0.0:6.1f}{-14.0:6.1f}{42.0:6.1f}{3.2:6.1f}",
    ]
    _write_sample(path, rows)
    assert is_wth_valid(path, end_year=2024)


def test_weather_validator_rejects_date_gaps(tmp_path):
    path = tmp_path / "TEST.WTH"
    _write_sample(path, [
        "2024001 12.0 10.0 1.0 0.0 0.0 40.0 3.0",
        "2024003 12.0 10.0 1.0 0.0 0.0 40.0 3.0",
    ])
    assert not is_wth_valid(path, end_year=2024)


def test_weather_validator_rejects_absolute_zero_temperature(tmp_path):
    path = tmp_path / "TEST.WTH"
    _write_sample(path, [
        f"{'2018001':>7}{0.0:6.1f}{-273.1:6.1f}{-273.1:6.1f}{0.0:6.1f}{-273.1:6.1f}{0.0:6.1f}{0.0:6.1f}"
    ])
    assert not is_wth_valid(path, end_year=2018)


def test_weather_validator_can_require_complete_core_forcing(tmp_path):
    path = tmp_path / "TEST.WTH"
    _write_sample(path, [
        "2024001 -99 10.0 1.0 0.0 0.0 40.0 -99",
        "2024002 12.0 10.0 1.0 0.0 0.0 40.0 -99",
    ])
    assert is_wth_valid(path, end_year=2024)
    assert not is_wth_valid(
        path, end_year=2024, required_columns=("SRAD", "TMAX", "TMIN", "RAIN")
    )


def test_weather_validator_can_require_all_agera5_forcing(tmp_path):
    path = tmp_path / "TEST.WTH"
    _write_sample(path, [
        "2024001 12.0 10.0 1.0 0.0 -99 -99 -99",
        "2024002 12.0 10.0 1.0 0.0 -99 -99 -99",
    ])
    core = ("SRAD", "TMAX", "TMIN", "RAIN")
    agera5 = core + ("TDEW", "RH2M", "WIND")

    assert is_wth_valid(path, end_year=2024, required_columns=core)
    assert not is_wth_valid(path, end_year=2024, required_columns=agera5)


def test_agera5_writer_defers_physical_validation_to_shared_validator(tmp_path):
    frame = pd.DataFrame({
        "DATE": ["2018001", "2018002"], "YEAR": [2018, 2018], "MM": [1, 1],
        "SRAD": [12.0, 12.0], "TMAX": [5.8, 10.0],
        "TMIN": [6.0, 2.0], "RAIN": [0.0, 0.0],
        "TDEW": [0.0, 0.0], "RH2M": [60.0, 60.0], "WIND": [3.0, 3.0],
    })
    path = Path(_write_wth(frame, "TEST", 33.7, -102.5, str(tmp_path)))

    assert path.exists()
    assert "   5.8   6.0" in path.read_text(encoding="utf-8")
    assert not is_wth_valid(
        path,
        end_year=2018,
        required_columns=("SRAD", "TMAX", "TMIN", "RAIN", "TDEW", "RH2M", "WIND"),
    )


def test_nasa_power_normalizes_json_null_to_numeric_missing(monkeypatch):
    class Response:
        def raise_for_status(self):
            return None

        def json(self):
            parameters = {
                name: {"19901103": (None if name == "ALLSKY_SFC_SW_DWN" else 1.0)}
                for name in _NASA_PARAMS
            }
            return {"properties": {"parameter": parameters}}

    monkeypatch.setattr(
        "dssatutils.weather_nasapower.requests.get",
        lambda *args, **kwargs: Response(),
    )

    frame = _fetch_nasa_power(34.57, -102.60, "19901103", "19901103", retries=1)

    assert frame.loc[0, "ALLSKY_SFC_SW_DWN"] == -99.0
    assert frame[_NASA_PARAMS].notna().all().all()
